lib/pbench/agent/tool_meister_client.py

```diff
@@ -0,0 +1,222 @@
+        channel_prefix=None,
+        logger=None,
+    ):
+        """Construct a Tool Meister "client" object, given the host and port of a
+        Redis server.  The caller can optionally provide a logger to be used.
+
+        This constructor does not contact the Redis server, it just records
+        the information for where that Redis server is.
+        """
+        assert (
+            redis_server is None and (redis_host is not None and redis_port is not None)
+        ) or (
+            redis_server is not None and (redis_host is None and redis_port is None)
+        ), "You must specify either a redis server object or a redis server host / port pair, but not both"
+
+        self.redis_server = redis_server
+        self.redis_host = redis_host
+        self.redis_port = redis_port
+
+        assert channel_prefix is not None, "You must specify a channel prefix"
+        self.to_client_channel = f"{channel_prefix}-{tm_channel_suffix_to_client}"
+        self.from_client_channel = f"{channel_prefix}-{tm_channel_suffix_from_client}"
+
+        if logger is None:
+            self.logger = logging.getLogger("tool-meister-client")
+        else:
+            self.logger = logger
+
+        self.pubsub = None
+
+    def __enter__(self):
+        """On context entry, setup the connection with the Redis server and discover
+        what Tool Meisters are out there.
+        """
+        if self.redis_server is None:
+            self.logger.debug("constructing Redis() object")
+            try:
+                self.redis_server = redis.Redis(
+                    host=self.redis_host, port=self.redis_port, db=0
+                )
+            except Exception as e:
+                self.logger.error(
+                    "Unable to connect to redis server, %s:%s: %s",
+                    self.redis_host,
+                    self.redis_port,
+                    e,
+                )
+                return 2
+            else:
+                self.logger.debug("constructed Redis() object")
+
+        # First setup our client channel for receiving action completion statuses.
+        self.logger.debug("pubsub")
+        self.pubsub = self.redis_server.pubsub()
+        self.logger.debug("subscribe %s", self.to_client_channel)
+        self.pubsub.subscribe(self.to_client_channel)
+        self.logger.debug("listen")
+        self.to_client_chan = self.pubsub.listen()
+        self.logger.debug("next")
+        resp = next(self.to_client_chan)
+        assert resp["type"] == "subscribe", f"Unexpected 'type': {resp!r}"
+        assert resp["pattern"] is None, f"Unexpected 'pattern': {resp!r}"
+        assert (
+            resp["channel"].decode("utf-8") == self.to_client_channel
+        ), f"Unexpected 'channel': {resp!r}"
+        assert resp["data"] == 1, f"Unexpected 'data': {resp!r}"
+
+        self.logger.debug("next success")
+        return self
+
+    def __exit__(self, *args):
+        if self.pubsub is not None:
```

Comment:
We should assert `self.pubsub` is not `None`, since the cleanup will only be called on context exit.